File: 270. Closest Binary Search Tree Value.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int closestValue(TreeNode* root, double target) {
        int result = 0;
        double prevdiff = numeric_limits<double>::max();
        while (root) {
            double sub = abs(root->val - target);
            if (sub < prevdiff) {
                result = root->val;
                prevdiff = sub;
            }
            if (target > root->val) root = root->right;
            else root = root->left;
        }
        return result;
    }
};

File: java/270. Closest Binary Search Tree Value.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// O(n)
public class Solution {
    public int closestValue(TreeNode root, double target) {
        Double diff = Double.MAX_VALUE;
        int result = 0;
        TreeNode cur = root;
        while (cur != null) {
            double abs = Math.abs(target - cur.val);
            if (abs < diff) {
                diff = abs;
                result = cur.val;
            }
            if (cur.val > target) {
                cur = cur.left;
            } else {
                cur = cur.right;
            }
        }
        return result;
    }
}
